src/passes/OptimizeInstructions.cpp

```diff
@@ -2815,6 +2818,31 @@ struct OptimizeInstructions
     return false;
   }
 
+  // (x << z) op (y << z)    ==>   (x op y) << z,   op = `|`, `&`, `^`
+  // (x >> z) op (y >> z)    ==>   (x op y) >> z,   op = `|`, `&`, `^`
+  // (x >>> z) op (y >>> z)  ==>   (x op y) >>> z,  op = `|`, `&`, `^`
+  Expression* combineBinaryWithShifts(Binary* curr) {
+    using namespace Abstract;
+    using namespace Match;
+    {
+      BinaryOp opLhs, opMid, opRhs;
+      Expression *x, *y, *z1, *z2;
+      if (matches(curr,
+                  binary(&opMid,
+                         binary(&opLhs, any(&x), any(&z1)),
+                         binary(&opRhs, any(&y), any(&z2)))) &&
+          (opLhs == opRhs) && hasAnyShift(opLhs) && hasAnyBitwise(opMid) &&
+          areConsecutiveInputsEqualAndFoldable(z1, z2)) {
```

Comment:
`z1` and `z2` are not consecutive, as `y` is in the middle.

Reply:
What if I add `canReorder(z1, y)` and preserve `areConsecutiveInputsEqualAndFoldable(z1, z2)`? `z1` and `z2` should be equal and without side effects to safely remove one of `z`.

Reply:
Yes, that sounds right. If we can reorder those two, then we end up the z's being consecutive.